library/dnet.c

```diff
@@ -393,6 +406,375 @@ static int dnet_iterator_callback_send(void *priv, void *data, uint64_t dsize)
+	 * this cleanup function is invoked, but since they were async,
+	 * client hasn't received ACK and there is no blocking
+	 * thread waiting for all requests to complete (like in iterator command)
+	 */
+	if (ctl->cmd.flags & DNET_FLAGS_NEED_ACK) {
+		dnet_send_ack(ctl->state, &ctl->cmd, err, 0);
+	}
+
+	pthread_cond_destroy(&ctl->write_wait);
+	pthread_mutex_destroy(&ctl->write_lock);
+
+	dnet_state_put(ctl->state);
+	free(ctl);
+
+	return err;
+}
+
+struct dnet_server_send_ctl *dnet_server_send_get(struct dnet_server_send_ctl *ctl)
+{
+	atomic_inc(&ctl->refcnt);
+	return ctl;
+}
+int dnet_server_send_put(struct dnet_server_send_ctl *ctl)
+{
+	if (atomic_dec_and_test(&ctl->refcnt))
+		return dnet_server_send_cleanup(ctl);
+
+	return 0;
+}
+
+static int dnet_server_send_sync(struct dnet_server_send_ctl *ctl)
+{
+	while (atomic_read(&ctl->writes_pending) > 0) {
+		pthread_mutex_lock(&ctl->write_lock);
+		pthread_cond_wait(&ctl->write_wait, &ctl->write_lock);
+		pthread_mutex_unlock(&ctl->write_lock);
+	}
+
+	return dnet_server_send_put(ctl);
+}
+
+/*
+ * Helper function which sends given data as WRITE command to remote groups
+ */
+int dnet_server_send_write(struct dnet_server_send_ctl *send,
+		struct dnet_iterator_response *re, uint64_t dsize,
+		int fd, uint64_t data_offset)
+{
+	struct dnet_net_state *client_state = send->state;
+	struct dnet_node *n = client_state->n;
+	struct dnet_io_control ctl;
+	struct dnet_session *s;
+	struct dnet_iterator_server_send_write_private *wp;
+	int err;
+
+	dnet_server_send_get(send);
+
+	/*
+	 * If need_exit is set - skips sending reply and return -EINTR to
+	 * interrupt execution of current iterator
+	 */
+	if (client_state->__need_exit) {
+		dnet_log(n, DNET_LOG_ERROR, "%s: Interrupting iterator because peer has been disconnected",
+				dnet_dump_id(&send->cmd.id));
+		err = -EINTR;
+		goto err_out_exit;
+	}
+
+	memset(&ctl, 0, sizeof(struct dnet_io_control));
+
+	memcpy(ctl.id.id, re->key.id, DNET_ID_SIZE);
+	ctl.id.group_id = 0; // will be rewritten in dnet_trans_create_send_all()
+
+	memcpy(ctl.io.id, re->key.id, DNET_ID_SIZE);
+	ctl.io.timestamp = re->timestamp;
+	ctl.io.user_flags = re->user_flags;
+	ctl.io.total_size = re->size;
+	ctl.io.size = re->size;
+	ctl.io.flags = DNET_IO_FLAGS_WRITE_NO_FILE_INFO;
+
+	// overwrite doesn't care whether remote key differs from local
+	// when this flag is not set, we only overwrite the same data or if there is no remote copy at all
+	if (!(send->iflags & DNET_IFLAGS_OVERWRITE))
+		ctl.io.flags |= DNET_IO_FLAGS_COMPARE_AND_SWAP;
+
+	// deliberately do not set DNET_FLAGS_NEED_ACK
+	// if WRITE command has failed, @dnet_process_cmd_with_backend_raw() will set this bit automatically,
+	// and will send acknowledge with error
+	//
+	// if there is no error, @dnet_file_info structure will be returned
+	ctl.cflags = 0;
+
+	ctl.fd = fd;
+	ctl.local_offset = data_offset;
+	ctl.cmd = DNET_CMD_WRITE;
```

Comment:
Again, that is the purpose of compare and swap write - fail it if checksum mismatch.
Yes, it will utilize the bandwidth. Yes, it will require reading data from disk.

The main purpose of iterator sending all of its data or some ranges of its data to remote nodes is to do this during recovery, in particular case when you do know that remote group is empty after crash or you are sure it doesn't contain content from this *local* node. It is a replacement for `rsync` and optionally for merge recovery, but in this case its main purpose is not only to send data but also to remove it locally.

If remote group is not empty, sending bulk load of writes will not be effective since most of the keys already exist.
That is why there is `session::server_send()` method which operates on per-key basis.